### PR TITLE
Update metadata download

### PR DESCRIPTION
**Pull request by zachthompson**
* Use new all/json endpoint so we don't need to make requests by IA type
* Request compressed response to reduce bandwidth

cc @jbarrett @jkv

### CODE REVIEW

File: lib/DDG/Meta/Data.pm [hunk 1]
```diff
@@ -36,91 +34,75 @@ unless(%ia_metadata){
...
-        my $c = $ENV{NO_METADATA_DOWNLOAD} ? 0 : mirror("https://duck.co/ia/repo/$json_endpt/json", $f);
-        my $json;
-        if($c == RC_OK || $c == RC_NOT_MODIFIED){
-            $json = io($f)->slurp; 
+    my $f = "$mdir/metadata.json";
+    unless($ENV{NO_METADATA_DOWNLOAD}){
+        my $ua = LWP::UserAgent->new;
+        $ua->timeout(5);
+        $ua->default_header('Accept-Encoding' => scalar HTTP::Message::decodable());
+        my $res = $ua->mirror("https://duck.co/ia/repo/all/json", $f);
```
Comment (nilnilnil) [inline lib/DDG/Meta/Data.pm]: What is the result of a partial read here?

File: lib/DDG/Meta/Data.pm [hunk 2]
```diff
@@ -36,91 +34,75 @@ unless(%ia_metadata){
...
+    }
 
-        # One metadata file for each repo with the following format
-        # { "<IA name>": {
-        #     "id": " " 
-        #     "signal" : " "
-        #     ....
-        # }
+    my $metadata = decode_json(io($f)->all);
+    next unless $metadata;
```
Comment (nilnilnil) [inline lib/DDG/Meta/Data.pm]: `next` what?

Comment (zachthompson) [inline lib/DDG/Meta/Data.pm]: Yep, loop is gone!

File: lib/DDG/Meta/Data.pm (hunk 1 above)
Comment (zachthompson) [inline lib/DDG/Meta/Data.pm]: No - https://metacpan.org/source/ETHER/libwww-perl-6.13/lib/LWP/UserAgent.pm#L929.  Download is saved to a temp file that is only swapped with the existing after several checks have passed.

Comment (nilnilnil) [inline lib/DDG/Meta/Data.pm]: Thx!

File: lib/DDG/Meta/Data.pm [hunk 3]
```diff
@@ -36,91 +34,74 @@ unless(%ia_metadata){
...
+            my $source = $ia->{src_id};
+            # by source number for fatheads
+            $ia_metadata{fathead_source}{$source} = $ia;
+            # by language for multi language wiki sources
+            # check that language is set since most fatheads don't have a language
+            if( my $lang = $ia->{src_options}{language}){
+                $ia_metadata{fathead_lang}{$lang} = $source;
             }
+            my $min_length = $ia->{src_options}{min_abstract_length};
+            $ia_metadata{fathead_min_length}{$source} = $min_length if $min_length;
```
Comment (russellholt) [inline lib/DDG/Meta/Data.pm]: is this all just indentation (83-104) ? hard to tell here on github.

Comment (nilnilnil) [inline lib/DDG/Meta/Data.pm]: @russellholt you can add `?w=1` to skip that BS: https://github.com/duckduckgo/duckduckgo/pull/150/files?w=1.

Comment (nilnilnil) [inline lib/DDG/Meta/Data.pm]: you can't comment in that mode though (which is lame).

Comment (zachthompson) [inline lib/DDG/Meta/Data.pm]: Mostly, yes, except we now need to check the repo to determine if it's a fathead.

File: lib/DDG/Meta/Data.pm [hunk 4]
```diff
@@ -36,91 +34,74 @@ unless(%ia_metadata){
...
-        next unless $metadata;
+    }
 
-        # One metadata file for each repo with the following format
-        # { "<IA name>": {
-        #     "id": " " 
-        #     "signal" : " "
-        #     ....
-        # }
+    my $metadata = decode_json(io($f)->all);
```
Comment (nilnilnil) [inline lib/DDG/Meta/Data.pm]: What made you determine we no longer need eval here? Is this whole thing worth putting in a `retry` block?

**Comment by nilnilnil** [comment]
:+1: I'm curious about the eval though. That has the potential to prevent processes from being up. Metadata needs to be excised from Constants, et al to make it less harmful when it fails.

File: lib/DDG/Meta/Data.pm (hunk 4 above)
Comment (zachthompson) [inline lib/DDG/Meta/Data.pm]: We want it to die here.  Before we were downloading four different files so only one of them was "necessary".

**Comment by kablamo** [comment]
@russellholt asked me to take a look.  I have 2 questions:

1. There is a $NO_METADATA_DOWNLOAD env variable.  What uses that?

2. When I load a module, I don't expect stuff to happen.  Generally I don't expect stuff to happen until I call a subroutine.  Its weird when I `use` something that uses something that uses something named Constants and end up downloading stuff from the internet.  To me this violates principle of least surprise.  In a web app I suppose this is not a big deal.  But of course, we use these libraries in non web situations as well.  Is it crazy to lazy load `%ia_metadata`?  ie encapsulate it in a getter method which caches the value when it first gets called?
```
%IA_METADATA
sub ia_metadata {
  return %IA_METADATA //= download_and_parse_metadata();
}
```

File: lib/DDG/Meta/Data.pm [hunk 5]
```diff
@@ -36,91 +34,74 @@ unless(%ia_metadata){
...
-        my $f = "$mdir/$iat.json";
-        my $c = $ENV{NO_METADATA_DOWNLOAD} ? 0 : mirror("https://duck.co/ia/repo/$json_endpt/json", $f);
-        my $json;
-        if($c == RC_OK || $c == RC_NOT_MODIFIED){
-            $json = io($f)->slurp; 
+    my $f = "$mdir/metadata.json";
+    unless($ENV{NO_METADATA_DOWNLOAD}){
+        my $ua = LWP::UserAgent->new;
+        $ua->timeout(5);
+        $ua->default_header('Accept-Encoding' => scalar HTTP::Message::decodable());
```
Comment (kablamo) [inline lib/DDG/Meta/Data.pm]: Interesting. I did not know about `HTTP::Message::decodable()`.

**Comment by nilnilnil** [comment]
If there is a way to lazy load it so that the first user request time is not effected across all of our processes on all of our servers, then I'm game. Otherwise, I think waiting avg 3.5 seconds for a metadata download & deserialize is not OK.

**Comment by zachthompson** [comment]
We want the download time to be eaten at restart I think, not on request.  

The env variable is just a way to opt-out of getting new metadata.  Anyone/thing can use it.  Nothing currently does but I've used it in the past when I need to manually modify the metadata to test.

**Comment by kablamo** [comment]
Meh.  Users, shmusers.  :)  ok you guys are right.  But I think I am too.  I guess what Catalyst and other web frameworks do is have an initialization phase which happens on startup.  Thats what we otta do.  Way outside the scope of this pr but here is my idea:  

When a web service is instantiated, an `initialize()` method could get run.  So in the psgi files we could something equivalent to this:
```
my $ads = DDG::Web::Ads->new();
$ads->initialize();
my $app = sub { $ads->run_psgi(@_) };

builder {
    $app
};
```
Not that exactly -- it shouldn't be so verbose.  But something like that.  In `initialize()` we would do this kind of startup downloading/initializing.   

Another thing a lot of frameworks do during this initialization phase is minify js and css and use an md5 of their contents to generate a unique filename to bust browser caches -- ie the stuff we do in filecache.  Just thinking out loud here.  Don't think that actually solves any of our filecache problems...

**Pull request merged**